### ISSUE TITLE
1.4 release

### ISSUE THREAD

**Issue by tjhei** [geodynamics/aspect#806]
I would like to start the release process for 1.4 and discuss what needs to happen before we release 1.4.

I know we were thinking to drop 8.2.1 support after 1.4 but I am thinking now that another option might be to just move the tester to 8.4.1 and require 8.4.1 for 1.4 and later. Thoughts?

Regarding the current open PRs, should we just delay most of them to after 1.4?

**Comment by bangerth** [comment]
There's no point in requiring 8.4.1 for ASPECT 1.4 -- we've always tested, and everything worked, with 8.2. So let's stick with this for the release and then switch the tester to 8.4.1 for any later development.

**Comment by bangerth** [comment]
We have 9 issues tagged for 1.4: https://github.com/geodynamics/aspect/milestones/1.4 . With the exception of #698 and #293, all others are documentation issues. None actually seem particularly critical, so I'd be ok with just bumping them to 1.5. What do you all think?

**Comment by tjhei** [comment]
> There's no point in requiring 8.4.1 for ASPECT 1.4

One problem is that we can not merge #790 (and I would like to have this in 1.4) if the tester still runs on 8.2.1 but I don't want to bump the tester to 8.4.1 without requiring 8.4.1 because we will likely break things on 8.2.1 soon after. Does that make sense?

> None actually seem particularly critical, so I'd be ok with just bumping them to 1.5.

agreed.

**Comment by bangerth** [comment]
On 04/05/2016 07:28 AM, Timo Heister wrote:
> One problem is that we can not merge #790
> <https://github.com/geodynamics/aspect/pull/790> (and I would like to
> have this in 1.4)

Why worry about this? It's a corner case that people are not likely 
going to encounter...

**Comment by gassmoeller** [comment]
I agree that we should start the release process in the near future (I am somewhat busy with EGU stuff for the next two weeks, but have more time afterwards). But what about #775 and #783 (they are both ready for review)? We are right in the middle of a transition of the assembly structure and there is quite some ugly code in there until #783 is merged (e.g. there are unused parameter warnings, and we unnecessarily compute and store an explicit_material_model that is never used).
As to the dependency to deal.II: I think for the release we should support 8.2.1, 8.3 and 8.4. Afterwards I would vote for dropping support for 8.2.1 (it will make the particles faster, and use less memory to finally use the active_cell_index), but I think it is too early to require 8.4. As to the tester: I feel best about running the tester with the oldest supported deal.II version (8.3 in that case), because I usually develop with a newer version and the tester then lets me know if I messed something up with the old version.
And I agree with Wolfgang about #790, I think it is a nice improvement and will prevent a few users from running into trouble, but it seems like a rare case, and I think the trouble we create be requiring everybody to migrate to deal 8.4 immediately just does not seem worth the benefit.

**Comment by tjhei** [comment]
> about #790 ... but it seems like a rare case

We already got at least one support email about it and it is trivial to trigger: ``mpirun -n 20 ./aspect convection_box.prm`` or something like that should be enough.

**Comment by bangerth** [comment]
On 04/05/2016 10:20 AM, Rene Gassmöller wrote:
> I agree that we should start the release process in the near future (I
> am somewhat busy with EGU stuff for the next two weeks, but have more
> time afterwards). But what about #775
> <https://github.com/geodynamics/aspect/pull/775> and #783
> <https://github.com/geodynamics/aspect/pull/783> (they are both ready
> for review)? We are right in the middle of a transition of the assembly
> structure and there is quite some ugly code in there until #783
> <https://github.com/geodynamics/aspect/pull/783> is merged (e.g. there
> are unused parameter warnings, and we unnecessarily compute and store an
> explicit_material_model that is never used).

Yes, I need to get these patches reviewed. Can you please mark them with 
1.4 as milestone?


> As to the dependency to deal.II: I think for the release we should
> support 8.2.1, 8.3 and 8.4. Afterwards I would vote for dropping support
> for 8.2.1 (it will make the particles faster, and use less memory to
> finally use the active_cell_index), but I think it is too early to
> require 8.4. As to the tester: I feel best about running the tester with
> the oldest supported deal.II version (8.3 in that case), because I
> usually develop with a newer version and the tester then lets me know if
> I messed something up with the old version.

If you force people to upgrade to any other version than a previously 
required one, then it makes sense to force an upgrade to the newest 
version available -- which would be 8.4.1 by then.

**Comment by tjhei** [comment]
> If you force people to upgrade to any other version than a previously
> required one, then it makes sense to force an upgrade to the newest
> version available -- which would be 8.4.1 by then.

Agreed. It makes our lives a lot easier and we can finally rip out all the backwards compatibility stuff.

**Comment by bangerth** [comment]
On 04/05/2016 11:25 AM, Timo Heister wrote:
>     about #790 <https://github.com/geodynamics/aspect/pull/790> ... but
>     it seems like a rare case
>
> We already got at least one support email about it and it is trivial to
> trigger: |mpirun -n 20 ./aspect convection_box.prm| or something like
> that should be enough.
I think we should have made that determination a couple of months ago. I 
vote to postpone this issue and take it as a reason to get the next 
release out more quickly instead.

**Comment by gassmoeller** [comment]
Ok, so we agree to require 8.4 after the release. 
I must say I changed my mind a bit  about #790, because today I ran into it myself. In a computation of around 3000 dofs per core the repartition algorithm decided to remove the area of one core completely (might be special for this setup). Could we split it into two separate PRs? We could merge the feature improvement now (it fixes things for 8.3 and 8.4, and replaces one cryptic error message by another one for 8.2), and leave a PR with the test open until after the 1.4 release?

**Comment by tjhei** [comment]
done, see #812

**Comment by tjhei** [comment]
Do we have any last minute changes before we can tag 1.4? Otherwise I would go ahead with it this week.

**Comment by gassmoeller** [comment]
No objections from my side. Let us know, if you need anything done by us.

**Comment by tjhei** [comment]
can anybody take a quick look at my first try at a changelog:
```
This release includes the following changes:
- Complete overhaul of the particle architecture.
- Large improvements to serial and parallel performance of particle handling.
- Support for time-dependent mesh refinement criteria.
- Flexible support for modifying the underlying FEM variables for the PDE.
- Optional DG support for temperature and compositional fields.
- More robust parallel file I/O.
- Support for tangential mesh velocity boundaries in free surface
  computations.
- Various fixes to free surface computations: checkpointing, better
  stabilization, crash fixes
- Support for traction boundary conditions.
- New Chunk Geometry model.
- New material property averaging options.
- Complete rewrite of the heating model infrastructure.
- Several new cookbooks.
- Several new postprocessors.
- Support for signals in various locations that allows plugins to
  inspect/manipulate things inside the core application.
- Several new mesh refinement plugins.
- Improved spherical interpolation of data used in the GPlates plugin. 
- Many other fixes and small improvements.
```

maybe reorder/add/remove some stuff?

**Comment by tjhei** [comment]
the release is tagged at https://github.com/geodynamics/aspect/releases/tag/v1.4.0

**Comment by bangerth** [comment]
The list looks good. Maybe somewhere add the changes to the assembly infrastructure somewhere? I suspect that people will like the traction boundary conditions, so maybe move that up a bit?

I think it's a good list. I assume you went through the changes.h file to make sure you got everything?

**Comment by bangerth** [comment]
How is it already a year (minus 3 days) since the last release? :-(

**Comment by ljhwang** [comment]
hackathons are powerful motivators for all!

**Comment by tjhei** [comment]
> Maybe somewhere add the changes to the assembly infrastructure somewhere?

Fixed. That was not in the changes.h.

> I suspect that people will like the traction boundary conditions, so maybe move that up a bit?

done.

> I assume you went through the changes.h file to make sure you got everything?

yes.

**Comment by tjhei** [comment]
Done! :fireworks:

**Comment by bangerth** [comment]
Thanks a lot, @tjhei -- I much appreciate getting off easy on this one!

**Comment by bangerth** [comment]
#884 is there to retroactively fix the pointed out lack of mention of the assembler rework in the 1.4 changelogs.
